### PR TITLE
DEPR: deprecate StataReader, StataWriter, genfromdta

### LINKED ISSUE #5761: IO: StataWriter and np.inf
TL;DR: is there a chance that pandas' StataWriter is doing a check that the statsmodels version should also be doing?

Looking at iolibs.foreign.StataReader and StataWriter and the updates pandas has made (their docstring specifically says it was based on Skipper's version here), I found that one of the statsmodels' tests raises when using the pandas version:

```
from statsmodels.iolib.tests.test_foreign import *
import pandas as pd

def test_missing_roundtrip():
    buf = BytesIO()
    dta = np.array([(np.nan, np.inf, "")],
                      dtype=[("double_miss", float), ("float_miss", np.float32),
                              ("string_miss", "a1")])
    writer = StataWriter(buf, dta)
    writer.write_file()
    buf.seek(0)
    dta = genfromdta(buf, missing_flt=np.nan)
    assert_(isnull(dta[0][0]))
    assert_(isnull(dta[0][1]))
    assert_(dta[0][2] == asbytes(""))

    dta = genfromdta(os.path.join(curdir, "results/data_missing.dta"),
            missing_flt=-999)
    assert_(np.all([dta[0][i] == -999 for i in range(5)]))


def test_missing_roundtrip_pd():
    buf = BytesIO()
    dta = np.array([(np.nan, np.inf, "")],
                      dtype=[("double_miss", float), ("float_miss", np.float32),
                              ("string_miss", "a1")])
    df = pd.DataFrame(dta)
    writer = StataWriter(buf, df)   # <-- raises
```

The relevant part of the traceback here reads:

```
  File "/usr/local/lib/python2.7/site-packages/pandas/io/stata.py", line 579, in _cast_to_stata_types
    'supported by Stata.'.format(col=col))
ValueError: Column float_miss has a maximum value of infinity which is outside the range supported by Stata.
```

Does statsmodels need to implement the same check?  If not, why not?

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #5761
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#writing-the-commit-message).

### CODE REVIEW

**Comment by bashtage** [comment]
@josef-pkt Any objections to starting this path?

File: statsmodels/iolib/foreign.py [hunk 1]
```diff
@@ -263,6 +265,11 @@ class StataReader(object):
             (-1.798e+308, +8.988e+307) }
 
     def __init__(self, fname, missing_values=False, encoding=None):
+        warnings.warn(
+            "StataReader is deprecated as of 0.10.0 and will be removed in a "
+            "future version.  Use pandas.read_stata instead.",
```
Comment (bashtage) [inline statsmodels/iolib/foreign.py]: Ideally something mentions both read_stata and StataReader, e.g., use pandas.read_state or pandas.io.stata.StataReader

File: statsmodels/iolib/foreign.py [hunk 2]
```diff
@@ -686,6 +693,10 @@ class StataWriter(object):
                        'd': 8.98846567431158e+307}
     def __init__(self, fname, data, convert_dates=None, encoding="latin-1",
                  byteorder=None):
+        warnings.warn(
+            "StataWriter is deprecated as of 0.10.0 and will be removed in a "
+            "future version.  Use pandas.DataFrame(data).to_stata instead.",
```
Comment (bashtage) [inline statsmodels/iolib/foreign.py]: Same comment: use pandas.DataFrame.to_stata or pandas.io.stata.StataWriter

Also, I think it should be pandas.DataFrame.to_stata, which is the correct function reference.

File: statsmodels/iolib/foreign.py [hunk 3]
```diff
@@ -956,6 +967,11 @@ def genfromdta(fname, missing_flt=-999., encoding=None, pandas=False,
         If convert_dates is True, then Stata formatted dates will be converted
         to datetime types according to the variable's format.
     """
+    warnings.warn(
+        "genfromdta is deprecated as of 0.10.0 and will be removed in a "
+        "future version.  Use pandas.read_stata instead.",
```
Comment (bashtage) [inline statsmodels/iolib/foreign.py]: Same here.

Comment (jbrockmendel) [reply statsmodels/iolib/foreign.py]: The other ones I'm clear on.  What are you saying should be added here?

Comment (bashtage) [reply statsmodels/iolib/foreign.py]: This is OK since it is the function interface.

**Comment by bashtage** [comment]
Random Fail.

**Pull request merged**